baking/src/tezos_baking/tezos_setup_wizard.py

```diff
@@ -405,21 +417,11 @@ def check_blockchain_data(self):
             return False
         return True
 
-    # Check the provider url and collect the most recent snapshot
-    # that is suited for the chosen history mode and network
-    def get_snapshot_metadata(self, name, json_url):
-        def hashes_comply(s1, s2):
-            return s1.startswith(s2) or s2.startswith(s1)
-
-        try:
-            snapshot_array = None
-            with urllib.request.urlopen(json_url) as url:
-                snapshot_array = json.load(url)["data"]
-            snapshot_array.sort(reverse=True, key=lambda x: x["block_height"])
-
-            node_version_hash = get_node_version_hash()
+    def extract_snapshot_metadata(self, snapshot_array):
+        from functools import reduce
 
-            snapshot_metadata = next(
+        def find_snapshot(pred):
+            return next(
                 filter(
```

Comment:
I suppose that there should be one major difference between the algorithm used here and the one used in Kiln.

Kiln is always releases with a stable Octez version and so we decided to filter out all the snapshots from Octez RCs.

Our packages however also get released for RCs, so we shouldn't do the same here.

IMO here we should filter out all snapshots from RCs only if we are using a stable Octez release and not otherwise.

Reply:
Why do you think we should filter out RC snapshots? When stable version just gets released, v18.0 for example, shouldn't latest snapshot for v18.0-rc1 be fine for it? (since there is no snapshots for v18.0 yet). But then I think we should also output node version information and snapshot version, so that user won't be very surprised if for some reason such import would fail.

Reply:
> Why do you think we should filter out RC snapshots?

Because, it being an RC, there's more chances that something is wrong with it.

Let's say that now we have a `v18.0` release and that a new `v18.1-rc1` comes out with the same bug as [tezos/#6318](https://gitlab.com/tezos/tezos/-/issues/6318) (which is what `v18.0-rc1` had), then with this new algorithm we'll have another broken release.

Small bugs like this are much more likely to happen with RCs and then get fixed before the next stable release.

> When stable version just gets released, v18.0 for example, shouldn't latest snapshot for v18.0-rc1 be fine for it?

Not necessarily, for example let's say that `v18.0` stopped supporting snapshot version `5` (which could have happened, since its snapshot version is `6`), then snapshots from `v18.0-rc1` would be incompatible (due to the same bug linked above).

Likewise in this case, and in general, we shouldn't make as many assumptions on RCs as we would for stable releases, because the former are intended to be tested more and (where necessary) fixed.

Reply:
As for handling RCs, I'd suggest doing the following (on the 2nd step):

* Reject all `rc` snapshots if user has a binary of stable version
* If user has an `rc` binary, then accept either
    a) snapshots with the same `major_version`, same `minor_version` and less `rc_version` (please note that we should reject stable snapshot with same major and minor version there too)
    b) **non-rc** snapshots with the same `major_version` and less `minor_version`
    
As for the `b)`, I think that there is no reason to care about `18.0-rc1` snapshots if we have `18.1-rc1` node because there definitely will be more fresh `18.0` snapshot. If there won't be such a snapshot (which is kinda unlikely), it's better to move to the next step anyway, since the current logic of handling the RCs is too complicated and error-prone (see my other comment)
    
@krendelhoff2 what do you think?